src/server/game/Movement/MotionMaster.cpp

```diff
@@ -727,131 +1106,120 @@ void MotionMaster::Mutate(MovementGenerator *m, MovementSlot slot)
+void MotionMaster::Delete(MovementGenerator* movement, bool active, bool movementInform)
 {
-    if (MovementGenerator* motion = GetMotionSlot(slot))
-    {
-        _slot[slot] = nullptr;
-        DirectDelete(motion);
-    }
-
-    while (!empty() && !top())
-        --_top;
+    TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::Delete: deleting generator (Priority: %u, Flags: %u, BaseUnitState: %u, Type: %u), owner: '%s'",
+        movement->Priority, movement->Flags, movement->BaseUnitState, movement->GetMovementGeneratorType(), _owner->GetGUID().ToString().c_str());
 
-    if (empty())
-        Initialize();
-    else if (NeedInitTop())
-        InitTop();
+    movement->Finalize(_owner, active, movementInform);
+    ClearBaseUnitState(movement);
+    MovementGeneratorPointerDeleter(movement);
 }
 
-void MotionMaster::DelayedClean(MovementSlot slot)
+void MotionMaster::DeleteDefault(bool active, bool movementInform)
 {
-    if (MovementGenerator* motion = GetMotionSlot(slot))
-    {
-        _slot[slot] = nullptr;
-        DelayedDelete(motion);
-    }
-
-    while (!empty() && !top())
-        --_top;
-}
-
-void MotionMaster::DirectExpire(bool reset)
-{
-    if (size() > 1)
-    {
-        MovementGenerator* curr = top();
-        pop();
-        DirectDelete(curr);
-    }
-
-    while (!empty() && !top())
-        --_top;
+    TC_LOG_DEBUG("movement.motionmaster", "MotionMaster::DeleteDefault: deleting generator (Priority: %u, Flags: %u, BaseUnitState: %u, Type: %u), owner: '%s'",
+        _defaultGenerator->Priority, _defaultGenerator->Flags, _defaultGenerator->BaseUnitState, _defaultGenerator->GetMovementGeneratorType(), _owner->GetGUID().ToString().c_str());
 
-    if (empty())
-        Initialize();
-    else if (NeedInitTop())
-        InitTop();
-    else if (reset)
-        top()->Reset(_owner);
+    _defaultGenerator->Finalize(_owner, active, movementInform);
+    _defaultGenerator = MovementGeneratorPointer(GetIdleMovementGenerator());
+    AddFlag(MOTIONMASTER_FLAG_STATIC_INITIALIZATION_PENDING);
 }
 
-void MotionMaster::DelayedExpire()
+void MotionMaster::AddBaseUnitState(MovementGenerator const* movement)
 {
-    if (size() > 1)
-    {
-        MovementGenerator* curr = top();
-        pop();
-        DelayedDelete(curr);
-    }
+    if (!movement || !movement->BaseUnitState)
+        return;
 
-    while (!empty() && !top())
-        --_top;
+    _baseUnitStatesMap.insert(std::make_pair(movement->BaseUnitState, movement));
+    _owner->AddUnitState(movement->BaseUnitState);
 }
 
-void MotionMaster::DirectDelete(MovementGenerator* curr)
+void MotionMaster::ClearBaseUnitState(MovementGenerator const* movement)
 {
-    if (IsStatic(curr))
+    if (!movement || !movement->BaseUnitState)
         return;
 
-    curr->Finalize(_owner);
-    delete curr;
-}
+    std::pair<MotionMasterUnitStatesContainer::iterator, MotionMasterUnitStatesContainer::iterator> bounds = _baseUnitStatesMap.equal_range(movement->BaseUnitState);
```

Comment:
Trinity::Containers::MultimapErasePair is designed exactly for what you are doing here, after erasing follow with
```
if (_baseUnitStatesMap.count(movement->BaseUnitState)
    _owner->ClearUnitState(movement->BaseUnitState);
```

Reply:
ow, thats a nice one 👍